### ISSUE TITLE
npm:discord.js hangs

### ISSUE THREAD

**Issue by ghaerdi** [denoland/deno#17391]
# Summary

I found a bug by trying to create a discord bot in Deno using the discord.js package from npm.
I ran the following code (replacing the empty string in the line 3):

```ts
import { Client, GatewayIntentBits, Events } from "npm:discord.js";

const TOKEN = "" // replace empty string with a discord bot's token;

const client = new Client({
  intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages],
});

client.on(Events.ClientReady, () => {
  console.log(`Logged in as ${client.user!.tag}!`);
});

client.on(Events.InteractionCreate, async (interaction) => {
  if (!interaction.isChatInputCommand()) return;

  if (interaction.commandName === "ping") {
    await interaction.reply("Pong!");
  }
});

client.login(TOKEN);
```
I got the following error in repl.it:
```
error: Uncaught TypeError: Deno.createHttpClient is not a function
    return Deno.createHttpClient({ caCerts });
                ^
    at HttpsClientRequest._createCustomClient (https://deno.land/std@0.167.0/node/https.ts:84:17)
    at async HttpsClientRequest._final (https://deno.land/std@0.167.0/node/http.ts:136:20)
```
But if I run the same code in my machine it doesn't output any error but it also doesn't initialize the bot, it just freeze in the last line.

The code should initialize the bot by changing bot status to online in discord, output in console "Logged in as {bot name and tag}" and if someone write "/ping" on a discord server (where is the bot) the bot should response with "pong". In node it runs as I described.

### Repl.it
Deno version: 1.28.3

### My Machine
Deno version: 1.29.2
OS: Arch Linux on WSL

**Comment by bartlomieju** [comment]
What is the command that you use to run this program?

**Comment by ghaerdi** [comment]
> What is the command that you use to run this program?

Here's my deno.jsonc file:
```json
{
  "tasks": {
    "dev": "deno run --allow-read --allow-env --allow-net --watch main.ts",
    "start": "deno run --allow-read --allow-env --allow-net --no-check main.ts"
  }
}
```
I tried with both tasks `deno task dev` and `deno task start`, also with `deno run -A main.ts` and in repl.it I used the default command when pressing the Start button with and without the `--no-check` flag: `deno run --lock=lock.json --lock-write --allow-all --no-check index.ts`.
I ran the same repl.it command in my machine and the output is the same: no errors and the bot doesn't run

**Comment by bartlomieju** [comment]
Does the problem go away if you add `--unstable` flag? `Deno.createHttpClient()` is an unstable API and is not available without that flag.

**Comment by ghaerdi** [comment]
Nothing happens, both in repl.it and in my machine, the returned promise from method `client.login` keeps in pending status.

**Comment by bartlomieju** [comment]
> Nothing happens, both in repl.it and in my machine, the returned promise from method `client.login` keeps in pending status.

But you don't get error about missing `Deno.createHttpClient()` anymore?

**Comment by ghaerdi** [comment]
You're right, no errors. But the bot doesn't initialize, in node it does.

**Comment by zookatron** [comment]
@ghaerdi You want to listen for debug events to see more details about what's going wrong, see the example below:

```
import { Client, Events, GatewayIntentBits } from "npm:discord.js";

const TOKEN = ""; // replace empty string with a discord bot's token;

const client = new Client({
  intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages],
});

client.on(Events.ClientReady, () => {
  console.log(`Logged in as ${client.user!.tag}!`);
});

client.on(Events.InteractionCreate, async (interaction) => {
  if (!interaction.isChatInputCommand()) return;

  if (interaction.commandName === "ping") {
    await interaction.reply("Pong!");
  }
});

client.on(Events.Debug, (message) => console.log("Debug:", message)); // Listen for debug messages

client.login(TOKEN);
```

@bartlomieju I investigated this issue and determined that this "hanging" is simply due to a persistent connection error that causes `discord.js` to retry the connection over and over forever. The connection error is due to the fact that the `https` deno compatibility shim does not properly support opening websocket connections, I've created an issue about this here: [https://github.com/denoland/deno/issues/17429](https://github.com/denoland/deno/issues/17429)

**Comment by evantill** [comment]
Thank you for the debug display
```js
client.on(Events.Debug, (message) => console.log("Debug:", message)); // Listen for debug messages
```
I was fighting against this "hanging" without understanding what was going on.

**Comment by LiberaTeMetuMortis** [comment]
Is there any updates about it?

**Comment by bartlomieju** [comment]
With https://github.com/denoland/deno/pull/19412 merged this is now supported. Will be released in v1.34.3 later this week.
